core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java

```diff
@@ -0,0 +1,188 @@
+            .specsById(table().specs())
+            .caseSensitive(isCaseSensitive())
+            .select(scanColumns())
+            .filterData(filter())
+            .filterManifestEntries(entry -> changelogSnapshotIds.contains(entry.snapshotId()))
+            .ignoreExisting();
+
+    if (shouldIgnoreResiduals()) {
+      manifestGroup = manifestGroup.ignoreResiduals();
+    }
+
+    if (newDataManifests.size() > 1 && shouldPlanWithExecutor()) {
+      manifestGroup = manifestGroup.planWith(planExecutor());
+    }
+
+    return manifestGroup.plan(new CreateDataFileChangeTasks(changelogSnapshots));
+  }
+
+  @Override
+  public CloseableIterable<ScanTaskGroup<ChangelogScanTask>> planTasks() {
+    return TableScanUtil.planTaskGroups(
+        planFiles(), targetSplitSize(), splitLookback(), splitOpenFileCost());
+  }
+
+  private Deque<Snapshot> changelogSnapshots(
+      Long fromSnapshotIdExclusive, long toSnapshotIdInclusive) {
+
+    Deque<Snapshot> changelogSnapshots = new ArrayDeque<>();
+
+    for (Snapshot snapshot :
+        SnapshotUtil.ancestorsBetween(table(), toSnapshotIdInclusive, fromSnapshotIdExclusive)) {
+
+      if (!snapshot.operation().equals(DataOperations.REPLACE)) {
+        if (snapshot.deleteManifests(table().io()).size() > 0) {
+          throw new UnsupportedOperationException(
+              "Delete files are currently not supported in changelog scans");
+        }
+
+        // insert at the beginning to have old snapshots first
+        changelogSnapshots.addFirst(snapshot);
+      }
+    }
+
+    return changelogSnapshots;
+  }
+
+  private static Map<Long, Integer> computeSnapshotOrdinals(Iterable<Snapshot> snapshots) {
+    Map<Long, Integer> snapshotOrdinals = Maps.newHashMap();
+
+    int ordinal = 0;
+
+    for (Snapshot snapshot : snapshots) {
+      snapshotOrdinals.put(snapshot.snapshotId(), ordinal++);
+    }
+
+    return snapshotOrdinals;
+  }
+
+  private static class CreateDataFileChangeTasks implements CreateTasksFunction<ChangelogScanTask> {
+    private static final DeleteFile[] NO_DELETES = new DeleteFile[0];
+
+    private final Map<Long, Integer> snapshotOrdinals;
+
+    CreateDataFileChangeTasks(Deque<Snapshot> snapshots) {
+      this.snapshotOrdinals = computeSnapshotOrdinals(snapshots);
+    }
+
+    @Override
+    public CloseableIterable<ChangelogScanTask> apply(
+        CloseableIterable<ManifestEntry<DataFile>> entries, TaskContext context) {
+
+      return CloseableIterable.transform(
+          entries,
+          entry -> {
+            Preconditions.checkArgument(
+                snapshotOrdinals.containsKey(entry.snapshotId()),
+                "Cannot determine ordinal for snapshot %s",
+                entry.snapshotId());
+
+            DataFile dataFile = entry.file().copy(context.shouldKeepStats());
+            int changeOrdinal = snapshotOrdinals.get(entry.snapshotId());
+            long commitSnapshotId = entry.snapshotId();
```

Comment:
nit: move this line above and can reuse the variable. current form is perfectly fine too

Reply:
Done.